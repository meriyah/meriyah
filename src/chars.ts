/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
