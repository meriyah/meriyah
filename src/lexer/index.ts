export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
