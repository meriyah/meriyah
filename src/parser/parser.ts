import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
