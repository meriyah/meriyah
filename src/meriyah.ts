import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
