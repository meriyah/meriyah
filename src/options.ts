import type * as ESTree from './estree';
import { pushComment, pushToken } from './parser/parser';
import { type Token } from './token';

/**
 * Function calls when semicolon inserted.
 */
type OnInsertedSemicolon = (pos: number) => any;

/**
 * Token process function.
 */
export type OnToken = (token: string, start: number, end: number, loc: ESTree.SourceLocation) => any;

/**
 * Comment process function.
 */
export type OnComment = (
  type: ESTree.CommentType,
  value: string,
  start: number,
  end: number,
  loc: ESTree.SourceLocation,
) => any;

/**
 * The parser options.
 */
export interface Options {
  // Allow module code
  module?: boolean;
  // Enable stage 3 support (ESNext)
  next?: boolean;
  // Enable start and end offsets to each node
  ranges?: boolean;
  // Enable web compatibility
  webcompat?: boolean;
  // Enable line/column location information to each node
  loc?: boolean;
  // Attach raw property to each literal and identifier node
  raw?: boolean;
  // Allow return in the global scope
  globalReturn?: boolean;
  // Enable implied strict mode
  impliedStrict?: boolean;
  // Enable non-standard parenthesized expression node
  preserveParens?: boolean;
  // Enable lexical binding and scope tracking
  lexical?: boolean;
  // Adds a source attribute in every nodeâ€™s loc object when the locations option is `true`
  source?: string;
  // Enable React JSX parsing
  jsx?: boolean;
  // Allows comment extraction. Accepts either a callback function or an array
  onComment?: ESTree.Comment[] | OnComment;
  // Allows detection of automatic semicolon insertion. Accepts a callback function that will be passed the character offset where the semicolon was inserted
  onInsertedSemicolon?: OnInsertedSemicolon;
  // Allows token extraction. Accepts either a callback function or an array
  onToken?: Token[] | OnToken;
}

export type NormalizedOptions = Omit<Options, 'onComment' | 'onToken'> & {
  onComment?: OnComment;
  onToken?: OnToken;
};

export function normalizeOptions(rawOptions: Options): NormalizedOptions {
  const options = { ...rawOptions } as NormalizedOptions;

  // Accepts either a callback function to be invoked or an array to collect comments (as the node is constructed)
  if (options.onComment) {
    options.onComment = Array.isArray(options.onComment) ? pushComment(options.onComment, options) : options.onComment;
  }

  // Accepts either a callback function to be invoked or an array to collect tokens
  if (options.onToken) {
    options.onToken = Array.isArray(options.onToken) ? pushToken(options.onToken, options) : options.onToken;
  }

  return options;
}
