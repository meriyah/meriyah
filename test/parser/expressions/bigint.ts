import * as t from 'node:assert/strict';
import { outdent } from 'outdent';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { pass } from '../../test-utils';

describe('Expressions - BigInt', () => {
  for (const arg of [
    '-1n',
    'const minus_one = BigInt(-1);',
    'x(30n, "foo", "bar");',
    '18757382984821n',
    '0b1111n',
    'a.b(c(0b1), 1n);',
    '340282366920938463463374607431768211456n',
    'a(3, 14n)',
    '431768211456n - 431768211456n',
    'class X { static async await(){} }',
    '(function x() { if (a === 0) return (1n / a) === (1n / b); })',
    '(function x() { for (var i = 0n; i < a.length; i++) { if (!x(a[i], b[i])) return false; } })',
    'var x = 2n ** 1000000000n;',
    '5n % 1n',
    '5n, 5n / 1n',
    'foo(5n, -5n / -1n);',
    '0n, 5n % 1n',
    '0n, -5n % -1n',
    '0n === 0n',
    'var x = 2n ** 31n - 1n;',
    'var x = kMaxInt - 32n - 2n;',
    'x = [2n ** 64n - 1n, 2n ** 64n - 2n, 4n, 3n, 2n, 1n, 0n]',
    'let x = 2n ** 30n - 1n;',
    'let x = 2n ** 30n;',
    '-4n',
    '-9007199254740992n',
    '(-9007199254740992n)',
    'a(-b.c - 3n) === -9007199254740994n',
    '1n + 1;',
    '1n + Object(1);',
    '1n + true;',
    '1 & Object(1n)',
    ' NaN | 1n;',
    'a(0b101n) << b(1n)',
    '0n << 128n === 0n',
    '0x246n << 127n === 0x12300000000000000000000000000000000n',
    '0b101n << 1n === 0b1010n',
    '0x246n << 0n === 0x246n',
    '0x123456789abcdef0fedcba9876543212345678n << -128n, 0x123456n',
    '-0x246n << 127n === -0x12300000000000000000000000000000000n',
    '-0x123456789abcdef0fedcba9876543212345678n << 64n, -0x123456789abcdef0fedcba98765432123456780000000000000000n',
    '-1n >>> -128n;',
    'a(1, 3n), -1n;',
    'a(async.await, 3n), -1n;',
    'a.b(BigInt(-9007199254740991), -9007199254740991n);',
    '(("0xf", 0), 0xfn);',
    'a(Number(0n), 0);',
    'let {} = 0n;',
    '() => 1n / 0n',
    'a(1n != false);',
    'a(1n != async);',
    '() => 1n ** -1n',
    '1n >= 1',
    'a(0n <= 1);',
    'a(1 <= 1n);',
    outdent`
      var data = [{
        a: 0x26ffcdbd233a53e7ca4612f2b02e1f2c1d885c3177e7n,
        r: 0x26ffcdbd233a53e7ca4612f2b02e1f2c1d885c3177e6n
      }, {
        a: 0xf2a29a35193377a223ef0d6d98db95eeb24a4165f288fd2b4an,
        r: 0xf2a29a35193377a223ef0d6d98db95eeb24a4165f288fd2b49n
      }, {
        a: 0x454d22e29e6104n,
        r: 0x454d22e29e6103n
      }, {
        a: -0xb00874640d30e6fce6bf79508378ed17e44dacb48a4200bce536cec462b3c2n,
        r: -0xb00874640d30e6fce6bf79508378ed17e44dacb48a4200bce536cec462b3c3n
      }, {
        a: 0x4c151a24d765249c2bab4a1915b24b80ae437417c5n,
        r: 0x4c151a24d765249c2bab4a1915b24b80ae437417c4n
      }, {
        a: -0xcbd476b1f9ca08ff820941n,
        r: -0xcbd476b1f9ca08ff820942n
      }, {
        a: -0xe848e5830fa1035322b39c2cdd031109ca8n,
        r: -0xe848e5830fa1035322b39c2cdd031109ca9n
      }, {
        a: -0x4d58c5e190f0ebac5bb36ca4d214069f69726c63a5n,
        r: -0x4d58c5e190f0ebac5bb36ca4d214069f69726c63a6n
      }, {
        a: 0x9b396n,
        r: 0x9b395n
      }, {
        a: 0x593921fe8b9d4906cn,
        r: 0x593921fe8b9d4906bn
      }, {
        a: -0xe127928c7cecd6e9ca94d98e858f9c76a0fccac62203aac7710cef1f9e352n,
        r: -0xe127928c7cecd6e9ca94d98e858f9c76a0fccac62203aac7710cef1f9e353n
      }, {
        a: 0xeb14cd952d06eb6fc613016f73b7339cbdd010n,
        r: 0xeb14cd952d06eb6fc613016f73b7339cbdd00fn
      }, {
        a: -0xfdeab6a3dbd603137f680413fecc9e1c80n,
        r: -0xfdeab6a3dbd603137f680413fecc9e1c81n
      }, {
        a: -0x7e9abbdfad170df2129dae8e15088a02b9ba99276a351a05n,
        r: -0x7e9abbdfad170df2129dae8e15088a02b9ba99276a351a06n
      }, {
        a: 0x7b98f57n,
        r: 0x7b98f56n
      }, {
        a: -0x919751deb470faa60d7c5c995c8bed72f9542d710fbbf1341n,
        r: -0x919751deb470faa60d7c5c995c8bed72f9542d710fbbf1342n
      }, {
        a: -0xc5541d89b118a88afdd187228440427c8a24f9d9bn,
        r: -0xc5541d89b118a88afdd187228440427c8a24f9d9cn
      }, {
        a: -0xe6c88a170595fn,
        r: -0xe6c88a1705960n
      }, {
        a: -0xa1ffbfa388c332804dc4dc973n,
        r: -0xa1ffbfa388c332804dc4dc974n
      }, {
        a: 0x67b768ce0c415127a77402861d1901dd7f60a8624ebea6ecafe03adc3cen,
        r: 0x67b768ce0c415127a77402861d1901dd7f60a8624ebea6ecafe03adc3cdn
      }];
    `,
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { webcompat: true });
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { sourceType: 'module' });
      });
    });
  }

  pass('Expressions - BigInt (pass)', [
    { code: '1n', options: { ranges: true } },
    { code: '1n + 2333333n', options: { ranges: true } },
    { code: '0xdead_BEEFn', options: { ranges: true, raw: true } },
  ]);
});
