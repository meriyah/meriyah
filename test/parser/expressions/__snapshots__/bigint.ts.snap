// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expressions - BigInt > Expressions - BigInt (pass) > 0xdead_BEEFn 1`] = `
{
  "body": [
    {
      "end": 12,
      "expression": {
        "bigint": "3735928559",
        "end": 12,
        "range": [
          0,
          12,
        ],
        "raw": "0xdead_BEEFn",
        "start": 0,
        "type": "Literal",
        "value": 3735928559n,
      },
      "range": [
        0,
        12,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 12,
  "range": [
    0,
    12,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - BigInt > Expressions - BigInt (pass) > 1n + 2333333n 1`] = `
{
  "body": [
    {
      "end": 13,
      "expression": {
        "end": 13,
        "left": {
          "bigint": "1",
          "end": 2,
          "range": [
            0,
            2,
          ],
          "start": 0,
          "type": "Literal",
          "value": 1n,
        },
        "operator": "+",
        "range": [
          0,
          13,
        ],
        "right": {
          "bigint": "2333333",
          "end": 13,
          "range": [
            5,
            13,
          ],
          "start": 5,
          "type": "Literal",
          "value": 2333333n,
        },
        "start": 0,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        13,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 13,
  "range": [
    0,
    13,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - BigInt > Expressions - BigInt (pass) > 1n 1`] = `
{
  "body": [
    {
      "end": 2,
      "expression": {
        "bigint": "1",
        "end": 2,
        "range": [
          0,
          2,
        ],
        "start": 0,
        "type": "Literal",
        "value": 1n,
      },
      "range": [
        0,
        2,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 2,
  "range": [
    0,
    2,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;
